contrib/platform/src/com/sun/jna/platform/win32/WbemcliUtil.java

```diff
@@ -0,0 +1,700 @@
+            if (o == null) {
+                return 0f;
+            } else if (vtTypeMap.get(property).equals(Variant.VT_R4)) {
+                return (Float) o;
+            }
+            throw new Wbemcli.WbemcliException(property.name() + " is not a Float type.", vtTypeMap.get(property));
+        }
+
+        /**
+         * Gets a Double value from the WmiResult. This is the return type when
+         * the WMI result is mapped to a VT_R8 (8-byte real) value.
+         * 
+         * @param property
+         *            The property (column) to fetch
+         * @param index
+         *            The index (row) to fetch
+         * @return The Double containing the specified value, or 0.0 if null
+         */
+        public Double getDouble(T property, int index) {
+            Object o = this.propertyMap.get(property).get(index);
+            if (o == null) {
+                return 0d;
+            } else if (vtTypeMap.get(property).equals(Variant.VT_R8)) {
+                return (Double) o;
+            }
+            throw new Wbemcli.WbemcliException(property.name() + " is not a Double type.", vtTypeMap.get(property));
+        }
+
+        /**
+         * Gets a value from the WmiResult, which may be null. Works with any
+         * return type. User must check for null and cast the result.
+         * 
+         * @param property
+         *            The property (column) to fetch
+         * @param index
+         *            The index (row) to fetch
+         * @return The Object containing the specified value, which may be null
+         */
+        public Object getValue(T property, int index) {
+            return this.propertyMap.get(property).get(index);
+        }
+
+        /**
+         * Adds a value to the WmiResult at the next index for that property
+         * 
+         * @param vtType
+         *            The Variant type of this object
+         * @param property
+         *            The property (column) to store
+         * @param o
+         *            The object to store
+         */
+        private void add(int vtType, T property, Object o) {
+            this.propertyMap.get(property).add(o);
+            if (vtType != Variant.VT_NULL && this.vtTypeMap.get(property).equals(Variant.VT_NULL)) {
+                this.vtTypeMap.put(property, vtType);
+            }
+        }
+
+        /**
+         * @return The number of results in each mapped list
+         */
+        public int getResultCount() {
+            return this.resultCount;
+        }
+
+        /**
+         * Increment the result count by one.
+         */
+        public void incrementResultCount() {
+            this.resultCount++;
+        }
+    }
+    
+    /**
+     * Private construtor for cleanup hook
+     */
+    private WbemcliUtil() {
+        Runtime.getRuntime().addShutdownHook(new Thread() {
+            @Override
+            public void run() {
+                OleAuto.INSTANCE.SysFreeString(WQL);
```

Comment:
I would not preallocated `WQL` and do the allocation in the query. At least I would measure if it really is a significant part of the call cost. This construction looks fragile.

Reply:
The allocation and release take an average of 33 microseconds for both calls.  Assuming (generously) I could kill one of them that'd save 16 microseconds per invocation.  So, yeah, not too significant. :)